### PR TITLE
Fetch service

### LINKED ISSUE #6: Code reusable fetch service
Add a reusable fetch service that can be used for making requests

### PR DESCRIPTION
**Pull request by ruddi10**
fixes #6

### CODE REVIEW

**Comment by ruddi10** [comment]
@aquibbaig  @muskankhedia would recommend reviewing this after the first one is merged. But if you wish you can have a look on last two commits

**Comment by muskankhedia** [comment]
@ruddi10 Please rebase it with the master branch.

**Comment by ruddi10** [comment]
> @ruddi10 Please rebase it with the master branch.

Done @muskankhedia

File: src/utils/useFetch.ts
Comment (muskankhedia) [inline src/utils/useFetch.ts:14]: add comments on the requirement of the function.

File: src/utils/useFetch.ts
Comment (muskankhedia) [inline src/utils/useFetch.ts:15]: have you tried testing the working of the function?

Comment (ruddi10) [reply src/utils/useFetch.ts:15]: Yes

File: src/utils/useFetch.ts
Comment (ruddi10) [reply src/utils/useFetch.ts:14]: @muskankhedia PTAL

**Comment by aquibbaig** [comment]
I think we can make this functionality a little simpler and more performant, lemme try to edit this and I will let you know, please continue to work on other stuff (I do not think this would be a major blocker in your work)

**Comment by ruddi10** [comment]
> I think we can make this functionality a little simpler and more performant, lemme try to edit this and I will let you know, please continue to work on other stuff (I do not think this would be a major blocker in your work)

So shall I make my machine selector component or ready for review or will it be reviewed after this one

**Comment by aquibbaig** [comment]
keep a local copy of this useFetch instance and use it. it needs simple changes but funcitonality will be the same

**Comment by aquibbaig** [comment]
@ruddi10 I rewrote this hook to make things easier when the project complexity increases. This snippet may look a little complicated but I think it is wise to use something like this. Let me know what you think about it.

```
import { useEffect, useReducer, useRef } from 'react';
import axios, { AxiosRequestConfig } from 'axios';

interface Response<T> {
  status: 'init' | 'fetching' | 'error' | 'fetched'
  data?: T | undefined
  error?: string | undefined
}

// used if you want to reference
// persistent storage
interface Cache<T> {
  [url: string]: T
}

type Action<T> =
  | { type: 'request' }
  | { type: 'success'; payload: T }
  | { type: 'failure'; payload: string }

const useFetch = <T = unknown>(
  url: string,
  options?: AxiosRequestConfig 
): Response<T> => {
  const cache = useRef<Cache<T>>({});
  const cancelRequest = useRef<boolean>(false);

  const initialState: Response<T> = {
    status: 'init',
    error: undefined,
    data: undefined,
  }
  
  // main function
  const fetchReducer = (state: Response<T>, action: Action<T>): Response<T> => {
    switch (action.type) {
      case 'request':
        return { ...initialState, status: 'fetching' }
      case 'success':
        return { ...initialState, status: 'fetched', data: action.payload }
      case 'failure':
        return { ...initialState, status: 'error', error: action.payload }
      default:
        return state
    }
  }

  // useReducer instead of use state could make the function
  // easier to debug across multiple usages.
  const [state, dispatch] = useReducer(fetchReducer, initialState);
  useEffect(() => {
    if (!url) {
      return
    }

    const fetchData = async () => {
      dispatch({ type: 'request' })

      if (cache.current[url]) {
        dispatch({ type: 'success', payload: cache.current[url] })
      } else {
        try {
          const response = await axios(url, options)
          cache.current[url] = response.data

          if (cancelRequest.current) return;

          dispatch({ type: 'success', payload: response.data })
        } catch (error: any) {
          // Let's keep error handling to another day,
         // hence the type for error is "any" here.
          if (cancelRequest.current) return;

          // server should return a message, else use a custom message such as "something went wrong"
          dispatch({ type: 'failure', payload: error.message })
        }
      }
    }

    fetchData();

    // runs when component is destroyed.
    return () => {
      cancelRequest.current = true
    }
  }, [url]);

  return state;
}

export default useFetch;
```

I think the data flow here follows a flow/redux style, which will be easier to debug and follow later on.

**Comment by aquibbaig** [comment]
The data flow here being 
Dispatch -> Action -> State -> View
as mentioned [here](https://facebook.github.io/flux/img/overview/flux-simple-f8-diagram-1300w.png)

**Comment by ruddi10** [comment]
> The data flow here being
> Dispatch -> Action -> State -> View
> as mentioned [here](https://facebook.github.io/flux/img/overview/flux-simple-f8-diagram-1300w.png)

So here instead of looking for isloaded = true we will be seeing the status right

**Comment by ruddi10** [comment]
> @ruddi10 I rewrote this hook to make things easier when the project complexity increases. This snippet may look a little complicated but I think it is wise to use something like this. Let me know what you think about it.
> 
> ```
> import { useEffect, useReducer, useRef } from 'react';
> import axios, { AxiosRequestConfig } from 'axios';
> 
> interface Response<T> {
>   status: 'init' | 'fetching' | 'error' | 'fetched'
>   data?: T | undefined
>   error?: string | undefined
> }
> 
> // used if you want to reference
> // persistent storage
> interface Cache<T> {
>   [url: string]: T
> }
> 
> type Action<T> =
>   | { type: 'request' }
>   | { type: 'success'; payload: T }
>   | { type: 'failure'; payload: string }
> 
> const useFetch = <T = unknown>(
>   url: string,
>   options?: AxiosRequestConfig 
> ): Response<T> => {
>   const cache = useRef<Cache<T>>({});
>   const cancelRequest = useRef<boolean>(false);
> 
>   const initialState: Response<T> = {
>     status: 'init',
>     error: undefined,
>     data: undefined,
>   }
>   
>   // main function
>   const fetchReducer = (state: Response<T>, action: Action<T>): Response<T> => {
>     switch (action.type) {
>       case 'request':
>         return { ...initialState, status: 'fetching' }
>       case 'success':
>         return { ...initialState, status: 'fetched', data: action.payload }
>       case 'failure':
>         return { ...initialState, status: 'error', error: action.payload }
>       default:
>         return state
>     }
>   }
> 
>   // useReducer instead of use state could make the function
>   // easier to debug across multiple usages.
>   const [state, dispatch] = useReducer(fetchReducer, initialState);
>   useEffect(() => {
>     if (!url) {
>       return
>     }
> 
>     const fetchData = async () => {
>       dispatch({ type: 'request' })
> 
>       if (cache.current[url]) {
>         dispatch({ type: 'success', payload: cache.current[url] })
>       } else {
>         try {
>           const response = await axios(url, options)
>           cache.current[url] = response.data
> 
>           if (cancelRequest.current) return;
> 
>           dispatch({ type: 'success', payload: response.data })
>         } catch (error: any) {
>           // Let's keep error handling to another day,
>          // hence the type for error is "any" here.
>           if (cancelRequest.current) return;
> 
>           // server should return a message, else use a custom message such as "something went wrong"
>           dispatch({ type: 'failure', payload: error.message })
>         }
>       }
>     }
> 
>     fetchData();
> 
>     // runs when component is destroyed.
>     return () => {
>       cancelRequest.current = true
>     }
>   }, [url]);
> 
>   return state;
> }
> 
> export default useFetch;
> ```
> 
> I think the data flow here follows a flux/redux style, which will be easier to debug and follow later on.

I think its a great implementation we can use this

**Comment by ruddi10** [comment]
@aquibbaig just did a small change firstly used axios.get instead of axios and checked if options are defined before using them in request did all this because it was causing some problem in my tests apart from that have checked its functioning and its working fine

**Comment by ruddi10** [comment]
@aquibbaig if this pr is ready and no more changes are there kindly inform me will squash these commits and push

**Comment by aquibbaig** [comment]
@ruddi10 the pr looks promising, lets wait for a final set of reviews and we can get it merged :)

**Comment by muskankhedia** [comment]
LGTM as well!

**Review by aquibbaig — APPROVED** [review]

**Pull request merged**
